src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,337 @@
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->AcceptedCommands(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Attributes(const ConcreteClusterPath & path,
+                                                   ReadOnlyBufferBuilder<DataModel::AttributeEntry> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->Attributes(path, out);
+}
+
+void CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged(const AttributePathParams & path)
+{
+    ChipLogDetail(DataManagement, "CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged");
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::AddEndpoint(EndpointProviderRegistration & registration)
+{
+    ReturnErrorOnFailure(mEndpointProviderRegistry.Register(registration));
+
+    // If the provider has already been started, we need to startup the clusters for this new endpoint.
+    if (mServerClusterContext.has_value())
+    {
+        EndpointProviderInterface * endpointProvider = registration.endpointProviderInterface;
+        VerifyOrReturnError(endpointProvider != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+        ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+        ReturnErrorOnFailure(endpointProvider->ServerClusterInterfaces(serverClusterBuilder));
+        auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+        for (auto * serverCluster : serverClusters)
+        {
+            VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+            CHIP_ERROR err = serverCluster->Startup(*mServerClusterContext);
+            if (err != CHIP_NO_ERROR)
+            {
+                // Log startup failure for the specific cluster
+                VerifyOrReturnError(!serverCluster->GetPaths().empty(), CHIP_ERROR_INCORRECT_STATE);
+                const ConcreteClusterPath path = serverCluster->GetPaths().front();
+                ChipLogError(DataManagement,
+                             "Cluster " ChipLogFormatMEI " on newly added Endpoint %u startup failed: %" CHIP_ERROR_FORMAT,
+                             ChipLogValueMEI(path.mClusterId), path.mEndpointId, err.Format());
+                // Continue trying to start other clusters, but return an error indicating partial failure.
```

Comment:
could we do this? Could we share the code with the other startup logic (since the same comment about incorrect state losing the startup error applies)?